### ISSUE TITLE
Mode collapse and batch normalization

### ISSUE THREAD

**Issue by BassyKuo** [meetshah1995/tf-3dgan#12]
Thanks for your implementation. I cloned this repo and saw the collapse during training [tf-3dgan/src/3dgan_mit_biasfree.py](https://github.com/meetshah1995/tf-3dgan/blob/master/src/3dgan_mit_biasfree.py), shown as below:
![image](https://user-images.githubusercontent.com/6426781/28839219-81b555ac-7725-11e7-8675-f4b477e91d99.png)

I tried to fix this problem by training lower dimension dataset (e.g. celebA) with the same model architecture. I found if I replace `tf.contrib.layers.batch_norm` with this [batchnormalize](https://github.com/yihui-he/GAN-MNIST/blob/master/mnist/model.py#L4) function, my training result will be better. For example in celebA: (with batch size=100)
![image](https://user-images.githubusercontent.com/6426781/28839148-4b92bb0e-7725-11e7-826f-6e6986739653.png)

Even I replaced with `tf.layers.batch_normalization`, the result was bad similar to `tf.contrib.layers.batch_norm`.

I don't know why the Tensorflow BN layer doesn't work. Do you have any idea?
Thank you so much.

**Comment by meetps** [comment]
I think in the latest API docs, one needs to update the batch estimates as every minibatch is passed. 

The docs say: 

```
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
  with tf.control_dependencies(update_ops):
    train_op = optimizer.minimize(loss)
```

So adding the above scope to these [lines](https://github.com/meetshah1995/tf-3dgan/blob/master/src/3dgan_mit_biasfree.py#L164-L167) will solve the problem.

**Comment by BassyKuo** [comment]
I add these lines and it really works in celebA!
![image](https://user-images.githubusercontent.com/6426781/28908429-2d0a00d4-7856-11e7-9da7-cad3db860392.png)
But it doesn't have an obvious change in 3D generating.
Thanks for your help.

**Comment by meetps** [comment]
Hey thanks for the update!

I think it isn't working for 3D inputs because the `tf.contrib.layers.batch_norm` normalizes over all axis except the last because the default format is `NHWC` [refer the inputs arguments](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm). I think changing the `tf.contrib.layers.batch_norm` to `tf.layers.batch_norm` will do the trick and normalization will only happen along the features axis.

Also could you make a PR as well with the updates that you made, it would be great to have them here :)

**Comment by BassyKuo** [comment]
Sorry for the late reply. Yes I used `tf.layers.batch_normalization` actually. Here is the training produce I got:
![image](https://user-images.githubusercontent.com/6426781/29002167-13a07788-7acf-11e7-9cbe-649f952c4a48.png)

I also sent my branch to you. Thank you!

**Comment by ghost** [comment]
@meetshah1995 @BassyKuo Thanks for your great work. I run  `tf-3dgan/src/3dgan_mit_biasfree.py`, but results are much worse than what you post on [Some sample generated chairs](https://github.com/meetshah1995/tf-3dgan) . And i couldn't get results as BassyKuo  got above too. 

Generated chairs during training(10000epochs totally):
![1](https://user-images.githubusercontent.com/13635264/30551626-a8025890-9ccd-11e7-8984-729d03a4345c.png)
![2](https://user-images.githubusercontent.com/13635264/30551650-b146a7bc-9ccd-11e7-9582-1af37db40c66.png)

Generated chairs using trained model biasfree_9500.cptk(9500epochs):
![4](https://user-images.githubusercontent.com/13635264/30551699-ea176cb6-9ccd-11e7-9260-ddfbc942ad29.png)
another try(different training process):
![5](https://user-images.githubusercontent.com/13635264/30551702-f2636d0c-9ccd-11e7-88b8-70505362919c.png)


Results are bad. And the network is unstable, it generates nothing but irregular clusters of points sometimes. Besides, we can see collapse as BassyKuo mentioned. It generates nearly same voxels. 

Any suggestion or trick to get results as good as yours? 
Thanks very much!
